slmsuite/hardware/slms/santec.py

```diff
@@ -258,15 +345,115 @@ def close(self):
 
     def _write_hw(self, phase):
         """See :meth:`.SLM._write_hw`."""
-        matrix = phase.astype(ctypes.c_uint16)
-
+        matrix = phase.astype(slm_funcs.USHORT)
         n_h, n_w = self.shape
 
         # Write to SLM
-        c = matrix.ctypes.data_as(ctypes.POINTER((ctypes.c_int16 * n_h) * n_w)).contents
-        ret = slm_funcs.SLM_Disp_Data(self.display_number, n_w, n_h, 0, c)
-        if ret != 0:
-            raise RuntimeError("SLM returned error {}.".format(ret))
+        c = matrix.ctypes.data_as(ctypes.POINTER((slm_funcs.USHORT * n_h) * n_w)).contents
+        Santec._parse_status(slm_funcs.SLM_Disp_Data(self.display_number, n_w, n_h, 0, c))
+
+    ### Additional Santec-specific functionality
+
+    def get_temperature(self):
+        """
+        Read the drive board and option board temperatures.
+
+        Returns
+        -------
+        (float, float)
+            Temperature in Celsius of the drive and option board
+        """
+        # Note that the Santec documentation suggests using signed 
+        # integers, but the header requests unsigned integers.
+        drive_temp = ctypes.c_uint32(0)
+        option_temp = ctypes.c_uint32(0)
+
+        Santec._parse_status(slm_funcs.SLM_Ctrl_ReadT(self.slm_number, drive_temp, option_temp))
+
+        return (drive_temp.value / 10., option_temp.value / 10.)
+
+    def get_error(self, raise_error=True):
+        """
+        Read the drive board and option board errors.
+
+        Returns
+        -------
+        list of str
+            List of errors.
+        """
+        drive_error = ctypes.c_uint32(0)
+        option_error = ctypes.c_uint32(0)
+
+        slm_funcs.SLM_Ctrl_ReadEDO(self.slm_number, drive_error, option_error)
+
+        # Check the resulting bitstrings for errors (0 ==> all good).
+        errors = []
+
+        for drive_error_bit in slm_funcs.SLM_DRIVEBOARD_ERROR.keys():
+            if drive_error.value & drive_error_bit:
+                errors.append(slm_funcs.SLM_DRIVEBOARD_ERROR[drive_error_bit])
+
+        for option_error_bit in slm_funcs.SLM_OPTIONBOARD_ERROR.keys():
+            if option_error.value & option_error_bit:
+                errors.append(slm_funcs.SLM_OPTIONBOARD_ERROR[option_error_bit])
+
+        if raise_error and len(errors) > 0:
+            raise RuntimeError("Santec error: " + ", ".join(["'" + err + "'" for err in errors]))
```

Comment:
Not worth redoing, but something to think about for future string building is that `"'" + err + "'"` creates 2 different strings (joins the first two then joins the combined first two with the third) and is somewhat less powerful than the format string construction, which lets you cast types and specify how many, e.g. decimals to display
[https://docs.python.org/3/library/string.html#formatstrings](https://docs.python.org/3/library/string.html#formatstrings)
So I would've done
"'{}'".format(err)
but they're basically the same and it doesn't really matter

Reply:
This code would have square brackets, but I guess it's easier to strip those.